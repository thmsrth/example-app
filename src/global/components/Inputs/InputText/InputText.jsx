import React from 'react';

import PropTypes from 'prop-types';
import validationKeysType from '../../Form/types/validationKeys';
import validationKeysDefaults from '../../Form/defaults/validationKeys';

import { bemCls } from '../../../utils/ClassNameHelpers';

const InputText = ({
  className, name, label, type, value, min, max, step, placeholder, htmlPattern,
  onChange, onBlur, rows, validationKeys, disabled,
}) => {
  const {
    required,
    validationAlert,
    warningAlert,
  } = validationKeys || InputText.defaultProps.validationKeys;

  switch (type) {
    case 'date':
    case 'text':
    case 'number':
    case 'password':
      return (
        <div className={`InputText ${className} ${
          bemCls(`InputText ${className}`, `__${name}`)} ${
          bemCls(`InputText ${className}`, '__required', required)} ${
          bemCls(`InputText ${className}`, '--alert', validationAlert)}`}
        >
          <label
            key={name}
            className={`${bemCls(`InputText ${className}`, '__label')} ${
              bemCls(`InputText ${className}`, `__${name}__label`)} ${
              bemCls(`InputText ${className}`, '__label--disabled', disabled)}`}
          >
            {`${label}${required ? '*' : ''}`}
            <div className={`${bemCls(`InputText ${className}`, '__alertMsg')} ${
              bemCls(`InputText ${className}`, `__${name}__alertMsg`)}`}
            >
              {validationAlert}
            </div>
            <input
              type={type}
              className={`${bemCls(`InputText ${className}`, '__input')} ${
                bemCls(`InputText ${className}`, `__${name}__input`)} ${
                bemCls(`InputText ${className}`, '__input--disabled', disabled)}`}
              name={name}
              value={!Number.isNaN(value) ? value : ''}
              min={type === 'text' ? undefined : min}
              max={type === 'text' ? undefined : max}
              step={type === 'number' ? step : undefined}
              pattern={htmlPattern || undefined}
              required={required}
              onChange={onChange}
              onBlur={onBlur}
              placeholder={placeholder}
              disabled={disabled}
            />
            <div className={`${bemCls(`InputText ${className}`, '__warningMsg')} ${
              bemCls(`InputText ${className}`, `__${name}__warningMsg`)}`}
            >
              {warningAlert}
            </div>
          </label>
        </div>
      );

    case 'textarea':
      return (
        <div className={`InputText ${className} ${
          bemCls(`InputText ${className}`, '__required', required)} ${
          bemCls(`InputText ${className}`, '--alert', validationAlert)}`}
        >
          <label
            key={name}
            className={`${bemCls(`InputText ${className}`, '__label')} ${
              bemCls(`InputText ${className}`, `__${name}__label`)} ${
              bemCls(`InputText ${className}`, '__label--disabled', disabled)}`}
          >
            {`${label}${required ? '*' : ''}`}
            <div className={`${bemCls(`InputText ${className}`, '__alertMsg')} ${
              bemCls(`InputText ${className}`, `__${name}__alertMsg`)}`}
            >
              {validationAlert}
            </div>
            <textarea
              className={`${bemCls(`InputText ${className}`, '__input')} ${
                bemCls(`InputText ${className}`, `__${name}__input`)} ${
                bemCls(`InputText ${className}`, '__input--disabled', disabled)}`}
              name={name}
              form={className}
              value={value}
              required={required}
              onChange={onChange}
              onBlur={onBlur}
              rows={rows}
              placeholder={placeholder}
            />
            <div className={`${bemCls(`InputText ${className}`, '__warningMsg')} ${
              bemCls(`InputText ${className}`, `__${name}__warningMsg`)}`}
            >
              {warningAlert}
            </div>
          </label>
        </div>
      );

    default:
      return null;
  }
};

export default InputText;

InputText.propTypes = {
  className: PropTypes.string,
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  type: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  min: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  max: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  step: PropTypes.string,
  htmlPattern: PropTypes.string,
  rows: PropTypes.string,
  validationKeys: validationKeysType,
  placeholder: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
};

InputText.defaultProps = {
  className: '',
  label: '',
  type: 'text',
  value: '',
  min: undefined,
  max: undefined,
  step: undefined,
  htmlPattern: '',
  rows: '5',
  validationKeys: validationKeysDefaults,
  placeholder: '',
  disabled: false,
  onChange: () => {},
  onBlur: () => {},
};
