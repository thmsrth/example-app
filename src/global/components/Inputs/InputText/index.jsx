import React, { Component } from 'react';

import PropTypes from 'prop-types';
import validationKeysType from '../../Form/types/validationKeys';
import validationKeysDefaults from '../../Form/defaults/validationKeys';

import InputText from './InputText';

class InputTextWrapper extends Component {
  constructor() {
    super();
    this.handleChange = this.handleChange.bind(this);
    this.handleBlur = this.handleBlur.bind(this);
  }

  handleChange(event) {
    const { name, onChange } = this.props;
    onChange(name, event.target.value);
  }

  handleBlur() {
    const {
      name, validationKeys, onBlur, type,
    } = this.props;
    const options = { type };

    onBlur(name, validationKeys, options);
  }

  render() {
    const { value, hidden, ...otherProps } = this.props;

    if (hidden) return null;

    return (
      <InputText
        {...otherProps}
        value={value}
        onChange={this.handleChange}
        onBlur={this.handleBlur}
      />
    );
  }
}

export default InputTextWrapper;

InputTextWrapper.propTypes = {
  name: PropTypes.string,
  type: PropTypes.string,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  validationKeys: validationKeysType,
  onChange: PropTypes.func,
  onBlur: PropTypes.func,
  hidden: PropTypes.bool,
};

InputTextWrapper.defaultProps = {
  name: '',
  type: 'text',
  value: '',
  validationKeys: validationKeysDefaults,
  onChange: () => {},
  onBlur: () => {},
  hidden: false,
};
