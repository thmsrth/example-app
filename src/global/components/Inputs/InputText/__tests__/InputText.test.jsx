import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import InputText from '../InputText';

const props = {
  className: 'MyInputText',
  name: 'mytext',
  label: 'My Input Label',
  type: 'text',
  value: 'My Text',
  validationKeys: {
    required: false,
    validationAlert: '',
    warningAlert: '',
  },
  onChange: () => {},
};

describe('InputText', () => {
  it('should render the component correctly with type text', () => {
    const tree = shallow(<InputText {...props} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with type number', () => {
    const tree = shallow(<InputText
      {...props}
      type="number"
      value="1"
      name="mynumber"
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component correctly with type text as a required input', () => {
    const tree = shallow(<InputText {...props} required />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component as a required input', () => {
    const tree = shallow(<InputText
      {...props}
      validationKeys={{ required: true }}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component with a validation alert message', () => {
    const tree = shallow(<InputText
      {...props}
      validationKeys={{ validationAlert: 'My validation alert message' }}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component with a warning message', () => {
    const tree = shallow(<InputText
      {...props}
      validationKeys={{ warningAlert: 'My warning message' }}
    />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
