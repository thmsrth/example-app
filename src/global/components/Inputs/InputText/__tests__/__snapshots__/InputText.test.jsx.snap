// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText should render the component as a required input 1`] = `
<div
  className="InputText MyInputText InputText__mytext MyInputText__mytext InputText__required MyInputText__required InputText--alert MyInputText--alert"
>
  <label
    className="InputText__label MyInputText__label InputText__mytext__label MyInputText__mytext__label "
    key="mytext"
  >
    My Input Label*
    <div
      className="InputText__alertMsg MyInputText__alertMsg InputText__mytext__alertMsg MyInputText__mytext__alertMsg"
    />
    <input
      className="InputText__input MyInputText__input InputText__mytext__input MyInputText__mytext__input "
      disabled={false}
      name="mytext"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      required={true}
      type="text"
      value="My Text"
    />
    <div
      className="InputText__warningMsg MyInputText__warningMsg InputText__mytext__warningMsg MyInputText__mytext__warningMsg"
    />
  </label>
</div>
`;

exports[`InputText should render the component correctly with type number 1`] = `
<div
  className="InputText MyInputText InputText__mynumber MyInputText__mynumber  "
>
  <label
    className="InputText__label MyInputText__label InputText__mynumber__label MyInputText__mynumber__label "
    key="mynumber"
  >
    My Input Label
    <div
      className="InputText__alertMsg MyInputText__alertMsg InputText__mynumber__alertMsg MyInputText__mynumber__alertMsg"
    />
    <input
      className="InputText__input MyInputText__input InputText__mynumber__input MyInputText__mynumber__input "
      disabled={false}
      name="mynumber"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      required={false}
      type="number"
      value="1"
    />
    <div
      className="InputText__warningMsg MyInputText__warningMsg InputText__mynumber__warningMsg MyInputText__mynumber__warningMsg"
    />
  </label>
</div>
`;

exports[`InputText should render the component correctly with type text 1`] = `
<div
  className="InputText MyInputText InputText__mytext MyInputText__mytext  "
>
  <label
    className="InputText__label MyInputText__label InputText__mytext__label MyInputText__mytext__label "
    key="mytext"
  >
    My Input Label
    <div
      className="InputText__alertMsg MyInputText__alertMsg InputText__mytext__alertMsg MyInputText__mytext__alertMsg"
    />
    <input
      className="InputText__input MyInputText__input InputText__mytext__input MyInputText__mytext__input "
      disabled={false}
      name="mytext"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      required={false}
      type="text"
      value="My Text"
    />
    <div
      className="InputText__warningMsg MyInputText__warningMsg InputText__mytext__warningMsg MyInputText__mytext__warningMsg"
    />
  </label>
</div>
`;

exports[`InputText should render the component correctly with type text as a required input 1`] = `
<div
  className="InputText MyInputText InputText__mytext MyInputText__mytext  "
>
  <label
    className="InputText__label MyInputText__label InputText__mytext__label MyInputText__mytext__label "
    key="mytext"
  >
    My Input Label
    <div
      className="InputText__alertMsg MyInputText__alertMsg InputText__mytext__alertMsg MyInputText__mytext__alertMsg"
    />
    <input
      className="InputText__input MyInputText__input InputText__mytext__input MyInputText__mytext__input "
      disabled={false}
      name="mytext"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      required={false}
      type="text"
      value="My Text"
    />
    <div
      className="InputText__warningMsg MyInputText__warningMsg InputText__mytext__warningMsg MyInputText__mytext__warningMsg"
    />
  </label>
</div>
`;

exports[`InputText should render the component with a validation alert message 1`] = `
<div
  className="InputText MyInputText InputText__mytext MyInputText__mytext InputText__required MyInputText__required InputText--alert MyInputText--alert"
>
  <label
    className="InputText__label MyInputText__label InputText__mytext__label MyInputText__mytext__label "
    key="mytext"
  >
    My Input Label
    <div
      className="InputText__alertMsg MyInputText__alertMsg InputText__mytext__alertMsg MyInputText__mytext__alertMsg"
    >
      My validation alert message
    </div>
    <input
      className="InputText__input MyInputText__input InputText__mytext__input MyInputText__mytext__input "
      disabled={false}
      name="mytext"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      type="text"
      value="My Text"
    />
    <div
      className="InputText__warningMsg MyInputText__warningMsg InputText__mytext__warningMsg MyInputText__mytext__warningMsg"
    />
  </label>
</div>
`;

exports[`InputText should render the component with a warning message 1`] = `
<div
  className="InputText MyInputText InputText__mytext MyInputText__mytext InputText__required MyInputText__required InputText--alert MyInputText--alert"
>
  <label
    className="InputText__label MyInputText__label InputText__mytext__label MyInputText__mytext__label "
    key="mytext"
  >
    My Input Label
    <div
      className="InputText__alertMsg MyInputText__alertMsg InputText__mytext__alertMsg MyInputText__mytext__alertMsg"
    />
    <input
      className="InputText__input MyInputText__input InputText__mytext__input MyInputText__mytext__input "
      disabled={false}
      name="mytext"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder=""
      type="text"
      value="My Text"
    />
    <div
      className="InputText__warningMsg MyInputText__warningMsg InputText__mytext__warningMsg MyInputText__mytext__warningMsg"
    >
      My warning message
    </div>
  </label>
</div>
`;
